components/Palettes/src/UIColor+MDCPalettable.m

```diff
@@ -0,0 +1,165 @@
+  for (size_t i = 0; i < numComponents + 1; ++i) {
+    if (fabs(components[i] - otherComponents[i]) > 0.001) {
+      return NO;
+    }
+  }
+  return YES;
+}
+
++ (MDCPalette *)paletteForColor:(UIColor *)color
+                    darkerColor:(UIColor * __autoreleasing*)darkerColor
+                   lighterColor:(UIColor * __autoreleasing*)lighterColor {
+  NSArray *palettes = @[MDCPalette.amberPalette,
+                        MDCPalette.blueGreyPalette,
+                        MDCPalette.bluePalette,
+                        MDCPalette.brownPalette,
+                        MDCPalette.cyanPalette,
+                        MDCPalette.deepOrangePalette,
+                        MDCPalette.deepPurplePalette,
+                        MDCPalette.greenPalette,
+                        MDCPalette.greyPalette,
+                        MDCPalette.indigoPalette,
+                        MDCPalette.lightBluePalette,
+                        MDCPalette.lightGreenPalette,
+                        MDCPalette.limePalette,
+                        MDCPalette.orangePalette,
+                        MDCPalette.pinkPalette,
+                        MDCPalette.purplePalette,
+                        MDCPalette.redPalette,
+                        MDCPalette.tealPalette,
+                        MDCPalette.yellowPalette];
+  for (MDCPalette *palette in palettes) {
+    NSArray *tints = @[
+                       palette.tint50,
+                       palette.tint100,
+                       palette.tint200,
+                       palette.tint300,
+                       palette.tint400,
+                       palette.tint500,
+                       palette.tint600,
+                       palette.tint700,
+                       palette.tint800,
+                       palette.tint900];
+    for (NSUInteger i = 0; i < tints.count; ++i) {
+      UIColor *tint = tints[i];
+      if (ColorEqualToColor(color, tint)) {
+        if (lighterColor && i > 0) {
+          *lighterColor = tints[i-1];
+        }
+        if (darkerColor && i < tints.count-1) {
+          *darkerColor = tints[i+1];
+        }
+        return palette;
+      }
+    }
+    if (palette.accent100) {
+      NSArray *accents = @[
+                           palette.accent100,
+                           palette.accent200,
+                           palette.accent400,
+                           palette.accent700];     
+      for (NSUInteger i = 0; i < accents.count; ++i) {
+        UIColor *accent = accents[i];
+        if (ColorEqualToColor(color, accent)) {
+          if (lighterColor && i > 0) {
+            *lighterColor = accents[i-1];
+          }
+          if (darkerColor && i < accents.count-1) {
+            *darkerColor = accents[i+1];
+          }
+          return palette;
+        }
+      }
+    }
+  }
+  return nil;
+}
+
+- (void)mdc_assignPaletteColors {
+  NSValue *flag = objc_getAssociatedObject(self, _cmd);
```

Comment:
I don't love that the object doesn't start with `is` or `has`. I know it's the method signature but maybe it should be a const? Is that possible with associated objects?